package/kernel/linux/modules/lib.mk

```diff
@@ -7,6 +7,79 @@
 
 LIB_MENU:=Libraries
 
+define KernelPackage/lib-crypto-blake2s
+  SUBMENU:=$(LIB_MENU)
+  TITLE:=BLAKE2s hash function library
+  KCONFIG:=CONFIG_CRYPTO_LIB_BLAKE2S
+  FILES_PLATFORM:=$(LINUX_DIR)/lib/crypto/libblake2s-generic.ko
+  FILES:=$(LINUX_DIR)/lib/crypto/libblake2s.ko $(FILES_PLATFORM)
+endef
+
+define KernelPackage/lib-crypto-blake2s/x86/64
+  KCONFIG+=CONFIG_CRYPTO_BLAKE2S_X86
+  FILES_PLATFORM:=$(LINUX_DIR)/arch/x86/crypto/blake2s-x86_64.ko
+endef
+
+$(eval $(call KernelPackage,lib-crypto-blake2s))
+
+
+define KernelPackage/lib-crypto-chacha20poly1305
+  SUBMENU:=$(LIB_MENU)
+  TITLE:=ChaCha20-Poly1305 AEAD support (8-byte nonce library version)
+  KCONFIG:=CONFIG_CRYPTO_LIB_CHACHA20POLY1305
+  FILES_PLATFORM:= \
+	$(LINUX_DIR)/lib/crypto/libpoly1305.ko \
+	$(LINUX_DIR)/lib/crypto/libchacha.ko
+  FILES:=$(LINUX_DIR)/lib/crypto/libchacha20poly1305.ko $(FILES_PLATFORM)
+endef
+
+define KernelPackage/lib-crypto-chacha20poly1305/x86/64
+  KCONFIG+= \
+	CONFIG_CRYPTO_CHACHA20_X86_64 \
+	CONFIG_CRYPTO_POLY1305_X86_64
+  FILES_PLATFORM:= \
+	$(LINUX_DIR)/arch/x86/crypto/poly1305-x86_64.ko \
+	$(LINUX_DIR)/arch/x86/crypto/chacha-x86_64.ko
+endef
+
+define KernelPackage/lib-crypto-chacha20poly1305/arm
+  KCONFIG+=CONFIG_CRYPTO_POLY1305_ARM
+  FILES_PLATFORM:=$(LINUX_DIR)/arch/arm/crypto/poly1305-arm.ko
+endef
+
+KernelPackage/lib-crypto-chacha20poly1305/kirkwood=$(KernelPackage/lib-crypto-chacha20poly1305/arm)
+
+define KernelPackage/lib-crypto-chacha20poly1305/arm-neon
+  $(call KernelPackage/lib-crypto-chacha20poly1305/arm)
+  KCONFIG+=CONFIG_CRYPTO_CHAHA20_NEON
+  FILES_PLATFORM+=$(LINUX_DIR)/arch/arm/crypto/chacha-neon.ko
+endef
+
+KernelPackage/lib-crypto-chacha20poly1305/imx6=$(KernelPackage/lib-crypto-chacha20poly1305/arm-neon)
+KernelPackage/lib-crypto-chacha20poly1305/ipq40xx=$(KernelPackage/lib-crypto-chacha20poly1305/arm-neon)
+KernelPackage/lib-crypto-chacha20poly1305/mvebu/cortexa9=$(KernelPackage/lib-crypto-chacha20poly1305/arm-neon)
+
+define KernelPackage/lib-crypto-chacha20poly1305/arm64-neon
+  $(call KernelPackage/lib-crypto-chacha20poly1305/arm-neon)
+  KCONFIG+=CONFIG_CRYPTO_POLY1305_NEON
+  FILES_PLATFORM+=$(LINUX_DIR)/arch/arm64/crypto/poly1305-neon.ko
+endef
+
+define KernelPackage/lib-crypto-chacha20poly1305/mips
```

Comment:
These should probably be split into separate chacha and poly ones, like you did for x86 and arm, rather than grouping these into one thing called chapoly.

Reply:
All target(board)-specific packages inherit from the generic kmod-lib-crypto-chacha20poly1305. The generic library is controlled with `CONFIG_CRYPTO_LIB_CHACHA20POLY1305`, but target-specific ones seem to be split into separate `CONFIG_` for chacha20 and poly1305. I could split it at the top level into kmod-lib-crypto-chacha20 and kmod-lib-crypto-poly1305 and then branch into targets from there if that is better. Is that what you are suggesting?

Reply:
Also, any reason why some chacha have 20 in the name and some don't?

Reply:
Take a look at the comment I just made on the root PR. This has a grouping I think would be sensible, given the symbols and dependencies.

ChaCha was originally specified with 20 rounds, but later people added reduced round variants of it. Linux uses ChaCha12 for a full disk encryption scheme, but uses ChaCha20 in the ChaCha20Poly1305 AEAD that WireGuard uses. Since the only difference between ChaCha20 and ChaCha12 is how many loop iterations are done, they share the same code in a module called ChaCha.